### PR TITLE
dragonball: add vhost-user connection management logic

### PR DESCRIPTION
**Pull request by adamqqqplay**
The vhost-user connection management logic will be used by
 the upcoming features: vhost-user-net, vhost-user-blk and
 vhost-user-fs.

Fixes: #8448

### CODE REVIEW

File: src/dragonball/src/dbs_utils/src/lib.rs [hunk 1]
```diff
@@ -2,6 +2,14 @@
 // Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+// Reexport timerfd public interfaces.
+pub mod timerfd {
+    pub use timerfd::*;
+}
+
+// Reexport vmm_sys_util public interfaces.
+pub use vmm_sys_util::{aio, errno, eventfd, fallocate, ioctl, tempdir, tempfile, terminal};
```
Comment (justxuewei) [inline src/dragonball/src/dbs_utils/src/lib.rs:11]: Are these re-export interfaces necessary? Shall we remove them?

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 2]
```diff
@@ -125,6 +125,32 @@ pub enum ActivateError {
     InvalidQueueConfig,
     #[error("IO: {0}.")]
     IOError(#[from] IOError),
+    #[error("VirtIo error")]
+    VirtIoError(Error),
+    #[error("VirtIo error")]
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:130]: Is error message correct?

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 3]
```diff
@@ -125,6 +125,32 @@ pub enum ActivateError {
...
+    EpollMgr(dbs_utils::epoll_manager::Error),
+    #[cfg(feature = "vhost")]
+    #[error("VirtIo error")]
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:133]: Is error message correct?

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 4]
```diff
@@ -125,6 +125,32 @@ pub enum ActivateError {
     InvalidQueueConfig,
     #[error("IO: {0}.")]
     IOError(#[from] IOError),
+    #[error("VirtIo error")]
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:128]:
```suggestion
    #[error("Virtio error")]
```

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 5]
```diff
@@ -125,6 +125,32 @@ pub enum ActivateError {
...
+    VirtIoError(Error),
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:129]: Ditto.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 6]
```diff
@@ -155,6 +181,9 @@ pub enum Error {
     /// Guest gave us a descriptor that was too big to use.
     #[error("descriptor length too big.")]
     DescriptorLengthTooBig,
+    /// Error from the epoll event manager
+    #[error("dbs_utils error: {0}.")]
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:185]:
```suggestion
    #[error("dbs_utils error: {0:?}.")]
```

Use it to print error stack.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 7]
```diff
@@ -229,6 +258,15 @@ pub enum Error {
     #[cfg(feature = "virtio-balloon")]
     #[error("Virtio-balloon error: {0}")]
     VirtioBalloonError(#[from] balloon::BalloonError),
+    
+    #[cfg(feature = "vhost")]
+    /// Error from the vhost subsystem
+    #[error("Vhost error: {0}")]
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:264]: Ditto.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs [hunk 8]
```diff
@@ -229,6 +258,15 @@ pub enum Error {
...
+    VhostError(vhost_rs::Error),
+    #[cfg(feature = "vhost")]
+    /// Error from the vhost user subsystem
+    #[error("Vhost-user error: {0}")]
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/lib.rs:268]: Ditto.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+    Error as VhostUserError, Listener as VhostUserListener, Master, VhostUserMaster,
+};
+use vhost_rs::{VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
+use virtio_queue::QueueT;
+use vm_memory::{
+    Address, GuestAddress, GuestAddressSpace, GuestMemory, GuestMemoryRegion, MemoryRegionAddress,
+};
+
+use super::super::super::device::VirtioDeviceConfig;
+use super::super::super::{Error as VirtIoError, Result as VirtIoResult};
```
Comment (justxuewei) [inline src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs:23]:
```suggestion
use super::super::super::{Error as VirtioError, Result as VirtioResult};
```

**Review by justxuewei — COMMENTED** [review]
Thanks @adamqqqplay, a few comments here.

File: src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs (hunk 9 above)
Comment (adamqqqplay) [reply src/dragonball/src/dbs_virtio_devices/src/vhost/vhost_user/connection.rs:23]: It looks like other files already use VirtIoError and VirtIoResult.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 2 above)
Comment (adamqqqplay) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:130]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 3 above)
Comment (adamqqqplay) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:133]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 6 above)
Comment (adamqqqplay) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:185]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 7 above)
Comment (adamqqqplay) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:264]: Fixed.

File: src/dragonball/src/dbs_virtio_devices/src/lib.rs (hunk 8 above)
Comment (adamqqqplay) [reply src/dragonball/src/dbs_virtio_devices/src/lib.rs:268]: Fixed.

**Review by justxuewei — APPROVED** [review]
Lgtm, thanks @adamqqqplay!

**Review by justxuewei — CHANGES_REQUESTED** [review]
Please replace `VirtIoXxx` with `VirtioXxx`, see: https://github.com/kata-containers/kata-containers/pull/8465.

**Comment by adamqqqplay** [comment]
> Please replace `VirtIoXxx` with `VirtioXxx`, see: #8465.

@justxuewei Fixed, PTAL.

**Review by justxuewei — APPROVED** [review]
Lgtm, thanks!

**Review by studychao — APPROVED** [review]

**Pull request merged**